atrium-api/src/types.rs

```diff
@@ -20,6 +20,46 @@ pub use integer::*;
 
 pub mod string;
 
+/// Trait for a collection of records that can be stored in a repository.
+///
+/// The records all have the same Lexicon schema.
+pub trait Collection: fmt::Debug {
```

Comment:
I don't know if `atrium_api::types` is the right place to put this trait. I'm happy to move it elsewhere if you have a preferred location.

Reply:
Hmmm... maybe it's time to start refactoring about `types.rs`. I initially wrote the `blob` definitions. here without considering that there would be so many more types-related codes.
From now on, would it be better to use the following structure?

```
types.rs
types/
    blob.rs
    cid_link_*.rs
    integer.rs
    string.rs
    ....
```

And how about making `types.rs` only contain the following? (RecordKey would be moved to string.rs, I think)

```rust
mod blob;
mod cid_link_...;
mod integer;
mod string;
...

pub use blob::*;
pub use cid_link_...::*;
pub use integer::*;
pub use string::*;
...
```

And `Collection` should be in ... `types/collection.rs`, or `types/traits.rs`, as a place to define traits? Or I think the traits could be kept in `types.rs` as a special thing.

In any case, I think it is OK to merge this pull-request and discuss it in another branches.

Reply:
When I added the `atrium_api::types::integer` submodule, my thinking was that we need a way to represent constraints on the single `integer` Lexicon type, but there aren't too many combinations, so these extra structs probably make sense in `atrium_api::types`. I put them in an `integer` submodule for easier local development.

When I added the `atrium_api::types::string` submodule, my thinking was that there are several well-defined `string` Lexicon formats, so the stuct for each string format make sense to be collected within `atrium_api::types::string` in the public API.

When creating this PR, my initial impulse was to put the `Collection` trait in `atrium_api::records`, but that is a generated module, so I put it in `atrium_api::types` instead.

My current thinking is that the lists of [types in the Data Model](https://atproto.com/specs/data-model#data-types) and/or the [top-level Lexicon types](https://atproto.com/specs/lexicon#overview-of-types) should map to either primitive Rust types (e.g. `bool`), or (if there is a 1:1 mapping) structs that live in `atrium_api::types`. Then any second-level Lexicon types (e.g. different formats of `string`) should map to structs that live either in a sub-module of `atrium_api::types`, or directly in `atrium_api::types` but with a name that indicates their Lexicon type (which is kind of what e.g. `BoundedU8` does, but it's less obvious).

My rationale here is that when importing a struct, `use atrium_api::types::{BoundedU8, string::Datetime, BlobRef}` make very clear what Lexicon types are involved (`integer`, `string`, and `blob` respectively). I'm also thinking that if we keep all type structs only exposed in `atrium_api::types`, that module is going to get rather unwieldy and difficult to navigate.

In that model, `record` is a primary Lexicon type, and the `Collection` trait is specifically a collection of `record` objects with the same Lexicon schema, so I'd imagine it would live in `atrium_api::types::record`.

> In any case, I think it is OK to merge this pull-request and discuss it in another branches.

Fine with me; this PR works for my use case as-is.